### ISSUE TITLE
IllegalReferenceCountException - "Failed to release a message: PooledSlicedByteBuf(freed)"

### ISSUE THREAD

**Issue by yamass** [rsocket/rsocket-java#857]
When I execute the following code, I get the error message mentioned in the title (and below). 

It only happens when not registering an error handler for a flux that actually throws an exception.

Anyway, this might cause severe problems (memory leak?), so I thought I would mention it.

## Expected Behavior
No IllegalReferenceCountException

## Actual Behavior
Exception and error log.

```
02-06-2020 19:07:25.228 [reactor-tcp-nio-2] WARN  i.n.u.ReferenceCountUtil - Failed to release a message: PooledSlicedByteBuf(freed)
io.netty.util.IllegalReferenceCountException: refCnt: 0, decrement: 1
	at io.netty.util.internal.ReferenceCountUpdater.toLiveRealRefCnt(ReferenceCountUpdater.java:74)
	at io.netty.util.internal.ReferenceCountUpdater.release(ReferenceCountUpdater.java:138)
	at io.netty.buffer.AbstractReferenceCountedByteBuf.release(AbstractReferenceCountedByteBuf.java:100)
	at io.netty.util.ReferenceCountUtil.release(ReferenceCountUtil.java:88)
	at io.netty.util.ReferenceCountUtil.safeRelease(ReferenceCountUtil.java:113)
	at reactor.netty.channel.ChannelOperationsHandler.channelRead(ChannelOperationsHandler.java:109)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:321)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:295)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:650)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:576)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:832)
```

## Steps to Reproduce

Sorry for this not being a real unit test, but you will get the idea.

```java
@Test
public void asdf() throws Exception {
	CloseableChannel server = RSocketServer.create((setup, sendingRSocket) -> Mono.just(new RSocket() {
		@Override
		public Flux<Payload> requestStream(Payload payload) {
			if (payload.getDataUtf8().equals("flux")) {
				return Flux.interval(Duration.ofMillis(1))
						.doOnNext(aLong -> System.out.println("Server-side doOnNext: " + aLong))
						.doOnCancel(() -> System.out.println("Server-side doOnCancel"))
						.map(aLong -> DefaultPayload.create("" + aLong));
			} else {
				return Flux.error(new RuntimeException("Some Exception for this test"));
			}
		}
	}))
			.bind(TcpServerTransport.create(TcpServer.create().host("localhost").port(7890)))
			.block();

	RSocket client = RSocketConnector.create()
			.connect(TcpClientTransport.create(TcpClient.create().host("localhost").port(7890)))
			.block();

	client.requestStream(DefaultPayload.create("exception"))
			.take(10)
			.subscribe(payload -> System.out.println(payload.getDataUtf8()));  // Note: no error handler!

	Thread.sleep(1000);
}
```

## Your Environment
* RSocket version(s) used: 1.0.0
* Other relevant libraries versions (eg. `netty`, ...): default rsocket dependencies
* Platform (eg. JVM version (`javar -version`) or Node version (`node --version`)):
openjdk 14.0.1 2020-04-14
OpenJDK Runtime Environment AdoptOpenJDK (build 14.0.1+7)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 14.0.1+7, mixed mode, sharing)
* OS and version (eg `uname -a`): MacOS

**Comment by OlegDokuka** [comment]
#853 related

**Comment by OlegDokuka** [comment]
@yamass thanks for reporting that. And your test is real UT (it compiles and runs!). 

Coming back to refcnt issue... The problem is not related to memory leaks. We have a different bug that may end up that buffer releases twice. I believe that the stacktrace was copied from your real app since the test is not reproducing it (also it may be because of the zero-copy decoder).

Any extra input is welcome.

Cheers,
Oleh

**Comment by yamass** [comment]
I also filed https://github.com/rsocket/rsocket-java/issues/858 . As mentioned, it might be related to this issue.

**Comment by OlegDokuka** [comment]
@yamass I'm afraid I will close this one as WONTFIX. This is the reactor specific issue that breaks all Reactive-Streams rules and I guess it should be fixed in rector rather than in RSocket. 

If it is not, 1.1 will be introducing specific operators that should minimize the reactor's impact.

**Comment by OlegDokuka** [comment]
A fix for #858 will be provided but we are not sure how the `ByteBuf` double release may be related to the error-handler issue since the unite test provided is not reproducing it.

Therefore, once #858 is fixed, please double check this one and if it is not reproducing anyhow or reproduces with another test, please provide extra input if you want us to continue working on this.

Cheers,
Oleh

**Comment by OlegDokuka** [comment]
@yamass 1.0.1 release fixed #858, and as I mentioned above, can you please double-check whether this issue can be reproduced (in particular double-release issue) again? And if so, can you please update repro steps?

Cheers,
Oleh

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by OlegDokuka** [comment]
Closed because of lack of information
